app/catalog_data.go

```diff
@@ -16,8 +21,71 @@ const (
 	systemLibraryURL    = "https://git.rancher.io/system-charts"
 	systemLibraryBranch = "master"
 	systemLibraryName   = "system-library"
+
+	defSystemChartVer = "manage.cattle.io/default-system-chart-version"
+)
+
+var (
+	sysDefaultBranch = os.Getenv("SYSTEM_CHART_DEFAULT")
 )
 
+
+// setSysDefaultBranch is meant to give a means of setting the system default branch without using
+// an environment variable and is purely meant for testing purposes.
+func setSysDefaultBranch(branch string) {
+	sysDefaultBranch = branch
+}
+
+// updateCatalogURL checks the default system-library version against the catalog actual version
+// and annotated version, then updates catalog's version and its annotation if an outdated version
+// is used.
+func updateCatalogURL(catalogClient v3.CatalogInterface) error {
+	oldDefaultURL := systemLibraryURL
+	oldDefaultBranch := systemLibraryBranch
+	defaultBranch := sysDefaultBranch
+	update := false
+
+	catalog, err := catalogClient.Get(systemLibraryName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Update if there is no annotation and url/branch are old defaults
+	if anno := catalog.Annotations[defSystemChartVer]; anno == "" {
+		if catalog.Spec.URL == oldDefaultURL && catalog.Spec.Branch == oldDefaultBranch {
+			catalog.Annotations = make(map[string]string)
+			update = true
+		}
+	} else {
+		annoMap := make(map[string]interface{})
+		err := json.Unmarshal([]byte(anno), &annoMap)
+		if err != nil {
+			return fmt.Errorf("%s annotation should be in JSON format: %v", defSystemChartVer, err)
```

Comment:
probably need to just log the error and say that system-chart catalog cannot be updated, rather than return an error, as that would 100% prevent rancher from starting up properly.